File: src/onscala/codility/stacksnqs/StoneWall.scala
package onscala.codility.stacksnqs

import org.scalatest.Assertions
import org.testng.Assert._
import org.testng.annotations.{DataProvider, Test}

/**
  * Site: HackerRank
  * Problem: Cyclic Rotations
  *
  * User: Luis Villaca
  * Date: 24/05/2020
  */
class StoneWall extends Assertions {

  def solution(h: Array[Int]): Int  = {
              if (h.isEmpty) 0
              else {
                var innerCache = List[Int]()
                var innerCount = 0

                for (h_item <- h) {
                  // 1 - descending - while new high < current cached entry,
                  //     iteratively ignore this entry and compare with previous one
                  while (!innerCache.isEmpty && innerCache.head > h_item) innerCache = innerCache.tail

                  // 2 - at the end of 1, cache may be empty, or current entry >= new high
                  //     add count and include new high in cache
                  if (innerCache.isEmpty || h_item > innerCache.head) {
                    innerCache = h_item :: innerCache
                    innerCount += 1
                  }

                  // 3 - plateau - while new high = current cached entry - do nothing
                }

                innerCount
              }

    }

  //Evaluating
  @DataProvider(name= "tests")
  def testData() = Array(
    Array[Any](Array(8, 8, 5, 7, 9, 8, 7, 4, 8), 7),
    Array[Any](Array(0), 1),
    Array[Any](Array(1, 1, 1), 1)
  )


  @Test(dataProvider = "tests")
  def verify(arrayInput: Array[Int], expectedOutput: Int): Unit = {
    val actualOutput = solution(arrayInput)
    assertEquals(actualOutput, expectedOutput)
  }

}

File: src/onjava/codility/stacksnqs/StoneWall.java
package onjava.codility.stacksnqs;


import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.Stack;

/**
 * Site: HackerRank
 * Problem: Stone Wall
 *
 * User: Luis Villaca
 * Date: 24/03/2020
 */
public class StoneWall {
    public int solution(int[] H) {
        Stack<Integer> stack= new Stack<>();

        int count=0;
        int index=0;

        // loop through array elements
        while(index<H.length) {

            // descending
            while (!stack.isEmpty() && H[index] < stack.peek()) {
                stack.pop();
                //do not increment index nor counter
                //instead, remove from cache and compare with previous entry
            }

            // ascending - add block
            if (stack.isEmpty() || H[index] > stack.peek()) {
                count++;
                stack.push(H[index]);
                index++;
                continue;
            } else {
                // if (H[index] == inStack) - plateau
                // do not increment counter, nor change cache
                index++;
            }
        }
        return count;
    }


    @DataProvider(name = "test")
    public Object [][] createData1() {
        return new Object [][] {
                new Object [] { new int [] { 8, 8, 5, 7, 9, 8, 7, 4, 8 }, 7 },
                new Object [] { new int [] { 0 }, 1 },
        };  //max
    }

    @Test(dataProvider = "test")
    public void verifySolution(int [] pA, int pExpectedMinNumberBlocks) {
        int actualMinNumberBlocksA = solution(pA);
        Assert.assertEquals(pExpectedMinNumberBlocks, actualMinNumberBlocksA);
    }
}
